# Select k nearest store by using Python Geopy 
## Module name: sel_nbrK_store
 1. Initialization   
 2. Retrive FAC_NBR, address from dw_bi_vw.dim_2_fac  

In [217]:
## Extract store address
"""
 Step0: Initialization 
  - Import packages/libraries from the file c_import.py
  - Setup database connection 
"""
prg_name = ""
path_code = "C:\\Users\\syue003\\wip_RecSys\\"
c_import  = path_code + "c_import.py"
c_setup_dbs_con = path_code + "c_setup_dbs_con.py"
c_timedte = path_code + "c_time_dte.py" 

exec(compile(open(c_import, 'rb').read(), c_import,  'exec'))
exec(compile(open(c_setup_dbs_con, 'rb').read(),c_setup_dbs_con, 'exec'))
exec(compile(open(c_timedte, 'rb').read(),c_timedte, 'exec'))
session, td_enginex = cf_setup_dbs_con(userName = 'syue003', passWord = 'Chungli#1')


t_engine teradata://syue003:Chungli#1@tqdpr02/temp_tables


In [ ]:

"""
 Define table name  
"""

#Load the transactional data into a Pandas dataframe  
query = """
    select fac_nbr,   
           ( trim(addr_line1_txt) || ' ' ||  trim( city_nm) || ' ' ||  trim( state_id) || ' '  || trim( zip5_id) ) 
           as address  
    from   dw_bi_vw.dim_2_fac   
    where  fac_nbr > 0
    group  by fac_nbr, address order by  fac_nbr
    """
df_str_address = pd.read_sql(query,session) 

In [ ]:
# Change data type of FAC_NBR to int64
df_str_address ['FAC_NBR']= df_str_address ['FAC_NBR'].astype(np.int64)

In [ ]:
"""
 Create store internal id
 
"""      
unq_fac_nbr = set(df_str_address.FAC_NBR.values)

facnbr_idx = {}
count = 0
for FAC_NBR in unq_fac_nbr:
    facnbr_idx[FAC_NBR] = count
    count += 1
df_str_address['facnbr_idx'] = df_str_address.apply(lambda row: facnbr_idx[row.FAC_NBR], axis=1)   

In [ ]:
"""
 Define RateLimiter object : geopy.extra.rate_limiter.RateLimiter
"""
locator = Nominatim(user_agent="myGeocoder")
geocode = RateLimiter(locator.geocode, min_delay_seconds=0.0)

In [ ]:
df = df_str_address

## Get geopy location and point



In [ ]:
"""
 Add a column "location" by applying "geocode" to df column "address" to get location
  - geocode will return geopy.Location.Location 
  - Very time consumming
"""

start_time = time.time()
df['location'] = df['address'].apply(geocode)
fnc_name = "Apply Geocode"
end_time = time.time()
cf_elapse_time (  start_time, end_time, "Function {0} completed.".format(fnc_name))

In [ ]:
"""
 Create column 'point' by applying point function to "location" 
 - point will return 'latitude', 'longitude', 'altitude'
   - will not extract  'altitude'
   - Define value of point as None if the value of location is None  
"""
start_time = time.time()
df['point'] = df['location'].apply(lambda loc: tuple(loc.point)[0:2] if loc else None)
end_time = time.time()
cf_elapse_time (  start_time, end_time, "Function {0} completed.".format("point"))

## Save dataframe to an excel file
 - It is time consuming to apply 'geocode' to get the geotype location
 - Save dataframe to an excel file for future use 

In [ ]:
df.to_excel('C:\SYUE\RecSys\df_str_addr_point.xlsx', index=False)

## Construct a distance matrix
 - Calculate distance between stores and save them in a  matrix 

   

In [ ]:
#Cal matrix of dis
df = pd.read_excel('C:\SYUE\RecSys\df_str_addr_point.xlsx')


In [222]:
df.head()

FAC_NBR                                        address  facnbr_idx  \
0        1            1340 PATRIOT BLVD GLENVIEW IL 60026           0   
1        2  201 SOUTH STEPHANIE STREET Henderson NV 89012           1   
2        3      14800 S E SUNNYSIDE RD Clackamas OR 97015           2   
3        4    3130 WEST CAREFREE HIGHWAY Phoenix AZ 85086           3   
4        5              11120 S. LAKE DR. RESTON VA 20191           4   

                                            location  \
0  Jewel-Osco, 1340, Patriot Boulevard, Glenview,...   
1  South Stephanie Street, Macdonald Ranch, Hende...   
2                                                NaN   
3  3130, West Carefree Highway, North Gateway, Ph...   
4                                                NaN   

                                   point  
0       (42.07856325, -87.8194154543438)  
1             (36.0324434, -115.0471777)  
2                                    NaN  
3  (33.7984748371083, -112.127522406678)  
4                                    NaN

### Construct a list l_point  to save the latitude, longitude to be ready for geopy.Point
- If the address is validate, the column point would be like  a string
  (latitude, longitude), e.g. (42.07856325, -87.8194154543438)
   - required to convert it to a string of 'latitude, longitude' 
      - A format accepted by geopy.Point
- If the address is invalidate,the value of point would be nan and the format would be float   
  - Will save ""  in the list 
- Use iterrows to iterate over DataFrame rows as (index, Series) pairs.
  - The columns of df are
    - FAC_NBR, address, facnbr_id, location, point
- l_point would be a list of strings with 'latitude, longitude' or empty string. 


In [ ]:
l_point = []
for index, row in df.iterrows(): 
    point = list(row)[4]  # Get VALUE FROM COLUMN "point"    
    if type(point) == float and math.isnan(point): 
       s_lat_lon = "" 
    else:      
       s_lat_lon = point[1:-1] #  Remove parenthesis
    l_point.append(s_lat_lon)

In [220]:
len (l_point)

3449

In [221]:
l_point[0:9]

['42.07856325, -87.8194154543438',
 '36.0324434, -115.0471777',
 '',
 '33.7984748371083, -112.127522406678',
 '',
 '',
 '46.000883, -112.519532',
 '39.5762384, -104.914441523727',
 '45.7727214646465, -111.185115020202']

## Calculate distance
- Convert the string  'latitude, longitude' in l_point to Point object and calculate the distance
  - If the p1 is 'nan' will assign 'nan' to the distance for all stores
  - If the p2 is 'nan' will assign 'nan' to the distance bwtween p1 and p2
  - Apply the geopy function distance.distance with parameter p1 and p2 to calculate the distane
    - Return the distance as mile
    
 ### Notes
  - Need to import math first, otherwise will get the mnessage of 'nan' is not defined;
    - nan cab be replaced by NaN

In [ ]:
distMatrix = []
l_point_x = l_point
l_nan = [float('nan') for i in range(len(l_point))]  # 

for s_lat_lon_1  in l_point:
    l_dis_x = []
    for s_lat_lon_2 in l_point_x:
        if s_lat_lon_1 is "":
           l_dis_x = l_nan
           continue
        else:
          if s_lat_lon_2 == "":           
             dist = float("nan") 
          else:
             p1 = Point(s_lat_lon_1)
             p2 = Point(s_lat_lon_2)
             dist = round(distance.distance(p1, p2).mi, 4)
          l_dis_x.append(dist)
    distMatrix.append(l_dis_x) 

In [223]:
len(distMatrix)

3449

In [ ]:
distMatrix[0: 2]

In [ ]:
## Build a dictionary 
   - Key  : FAC_NBR
   - value: Inner store id      

In [169]:
keys = df.facnbr_idx.to_list()
val_fac_nbr = df.FAC_NBR.to_list()
d_fac_nbr = dict(zip(keys, val_fac_nbr))
val_addr = df.address.to_list()
d_fac_nbr_addr = dict(zip(keys, val_addr))

In [225]:
d_fac_nbr[0]

1

## Get k nearest stores
- Build an array to get all the distance  
   -  store_dist: a list to have all the inner store id, the distance related to the target store ssid
      -  store_dist will pass to heapq.nsmallest to get K nearst strores
         - heapq.nsmallest is a Python method to sore a dictionary by value
      - innerID: inner store id
         - kNeighbors is a list of tuple with (innerID, dist), e.g.  
            [(iid0, dist0), (iid1, dist1) .... (iid9, dist9)] 
            - iid0 would have the smallest distance and is close to siid most
 

In [212]:
l_knbr=[]
for siid in range(len(df)):  
   
    distRow = distMatrix [siid]
    
    store_dist = []        
    for innerID, dist in enumerate(distRow):
        if (innerID != siid and math.isnan(float(dist) )!= True):
            store_dist.append( (innerID, dist) )  
  
    if len(store_dist) != 0:        
       kNeighbors = heapq.nsmallest(K, store_dist, key=lambda strDist: strDist[1]) 
       knbr_x     = list (itertools.chain(*kNeighbors)) 
       
       knbr_id    = knbr_x[::2]       # Get internal store id from knbr_x (:2 would skip the dist)
       knbr_dist  = knbr_x[1::2]      # get distance 
       knbr_strid = [d_fac_nbr[x] for x in knbr_id]      # Convert internal id to raw id  
       knbr_addr  = [d_fac_nbr_addr[x] for x in knbr_id] # Get store address
          
       
       for i in range(K):
           rid = d_fac_nbr[siid]
           dist_rnk = i + 1
           l_knbr.append([rid, knbr_strid[i], knbr_dist[i], dist_rnk, knbr_addr[i]] )    
      

## Create a daframe from l_knbr

In [214]:

df_knbr = pd.DataFrame(l_knbr, columns =['FAC_NBR', 'knbr_strid', 'knbr_dist', 'dist_rnk', 'knbr_addr']) 

In [226]:
df_knbr.head(20)

FAC_NBR  knbr_strid  knbr_dist  dist_rnk  \
0         1        1017   1.082132         1   
1         1        1055   5.156653         2   
2         1        1031   5.229062         3   
3         1        1104   5.315645         4   
4         1        1168   5.497983         5   
5         1        1116   5.915131         6   
6         1        1140   6.065405         7   
7         1        1035   6.368034         8   
8         1        1096   8.053811         9   
9         1          74   8.491886        10   
10        2          14   4.878778         1   
11        2        1062   8.201255         2   
12        2         335   8.903334         3   
13        2          93   9.609410         4   
14        2          45   9.770761         5   
15        2         277  11.071911         6   
16        2          89  11.628702         7   
17        2          46  14.045412         8   
18        2          13  16.060189         9   
19        2          61  16.837616        10   

                                            knbr_addr  
0                    1414 LOOP 336 W. CONROE TX 77304  
1       7720 EAST HIGHWAY 69 PRESCOTT VALLEY AZ 86314  
2          2931 CENTRAL CITY BLVD. GALVESTON TX 77551  
3   13401 WASHINGTON BOULEVARD Marina del Rey CA 9...  
4      3443 SOUTH SEPULVEDA BLVD Los Angeles CA 90034  
5          1050 KEN PRATT BOULEVARD LONGMONT CO 80501  
6                    1150 W. MAPLE MUNDELEIN IL 60060  
7                     9503 JONES RD. HOUSTON TX 77065  
8            801 KENILWORTH BLVD. KENILWORTH NJ 07033  
9       7000 W. FOREST PRESERVE DR. NORRIDGE IL 60176  
10               575 COLLEGE DRIVE Henderson NV 89015  
11                 4754 42ND AVE S W SEATTLE WA 98116  
12         2825 W KENNEWICK AVENUE KENNEWICK WA 99336  
13          1008 NEVADA HIGHWAY Boulder City NV 89005  
14  1760 EAST CHARLESTON BOULEVARD Las Vegas NV 89104  
15             2125 WEST HIGHWAY 70 THATCHER AZ 85552  
16           6885 E LAKE MEAD BLVD Las Vegas NV 89156  
17           1001 S. Rainbow Blvd. Las Vegas NV 89145  
18     2550 SOUTH FORT APACHE ROAD Las Vegas NV 89117  
19            4850 WEST CRAIG ROAD Las Vegas NV 89032

## load the dataframe to a table

In [219]:
    fnc_name = 'Load table'
    start_time = time.time()
    tbl_nm = "rs_nearst_store"
    
    df_knbr.to_sql(con=td_enginex, name=tbl_nm, if_exists='replace', index = False)
    end_time = time.time()
    cf_elapse_time (  start_time, end_time, "Function {0} to load table {1} completed.".format(fnc_name, tbl_nm))

    

 Function Load table to load table rs_nearst_store completed. It took 228.339831 seconds - 0hh:3mm:48ss.
 start time: Oct 08 2019 12:51:55  end time:  Oct 08 2019 12:55:44


In [227]:
df.location.isnull().sum()

2637